cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeServiceRoutingHandler.java

```diff
@@ -182,10 +183,11 @@ public void handleError(Throwable throwable) {
    * Opens a {@link HttpURLConnection} to the given service for the given program run.
    *
    * @throws BadRequestException if the request for service routing is not valid
+   * @throws GoneException if the run already finished
    */
   private HttpURLConnection openConnection(HttpRequest request, String namespace, String app,
                                            String version, String programType, String program, String run,
-                                           String service) throws BadRequestException {
+                                           String service) throws BadRequestException, GoneException {
```

Comment:
why do we need GoneException here?

Reply:
We check with RequestValidatior and it may throw a GoneException if the run id already completed